### PR TITLE
RTTS-9: Presentational List component

### PR DESCRIPTION
**Pull request by Sumragen**
#9

### CODE REVIEW

File: src/App/App.tsx [hunk 1]
```diff
@@ -42,41 +41,27 @@ const tKey = createTKey("app");
...
-      </ListItem>
-    </List>
-    <Fab color="primary" aria-label="add" className={classes.fab}>
-      <Add />
-    </Fab>
-  </div>
-));
+)(({ classes, t }: Props) => {
+  const timeEntries: TimeEntryI[] = JSON.parse(`[{"aspect":"aspect 1","project":"project 1","type":"type 1","task":"task 1"},{"aspect":"aspect 2","project":"project 2","type":"type 2","task":"task 2"},{"aspect":"aspect 3","project":"project 3","type":"type 3","task":"task 3"},{"aspect":"aspect 4","project":"project 4","type":"type 4","task":"task 4"},{"aspect":"aspect 5","project":"project 5","type":"type 5","task":"task 5"},{"aspect":"aspect 6","project":"project 6","type":"type 6","task":"task 6"}]`);
+  // const timeEntries: TimeEntryI[] = [];
```
Comment (dezoxel) [inline src/App/App.tsx]: Eventual PR for merging should not have a commented out code, the only exception is adding `// todo:`, explaining why it was commented. So please, add todo or remove the comment.

File: src/App/App.tsx [hunk 2]
```diff
@@ -42,41 +41,27 @@ const tKey = createTKey("app");
...
+       <CssBaseline />
+       <AppBar position="static">
+         <Toolbar>
+           <IconButton edge="start" color="inherit" aria-label="menu">
+             <Menu />
+           </IconButton>
+           <Typography variant="h6">{t(tKey("title"))}</Typography>
+         </Toolbar>
+       </AppBar>
+       <TimeEntries entries={timeEntries}/>
```
Comment (dezoxel) [inline src/App/App.tsx]: Minor, but I would recommend to always add an empty space between `}` and `/>`

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import React, { ComponentType } from "react";
+import { compose } from "recompose";
+import { createStyles, Theme, withStyles, WithStyles } from "@material-ui/core";
+import { WithTranslation, withTranslation } from "react-i18next";
+import { TimeEntryI } from "../TimeEntry.model";
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.tsx]: Please divide imports with an empty line. Also, I would suggest adding style guide rule about import grouping.

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+interface Props
+  extends CompletedTimeEntriesProps,
+    WithStyles<typeof styles>,
+    WithTranslation {}
+
+const styles = ({ palette }: Theme) =>
+  createStyles({
+    root: {
+      color: palette.primary.main
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.tsx]: I don't think that you need to specify a text color here. I understand that other parts contained such code, but it was an example. So in eventual code we have to remove any unnecessary styles. As I can see here, you don't need styles at all, so please remove `styles` term, `withStyles` HOC and corresponding typings.

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    }
+  });
+
+// prettier-ignore
+export const CompletedTimeEntries: ComponentType<CompletedTimeEntriesProps> = compose<Props, CompletedTimeEntriesProps>(
+  withTranslation(),
+  withStyles(styles, { withTheme: true, name: "RTTSCompletedTimeEntries" })
+)(({ classes, entries}) => (
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.tsx]: Empty space after `entries` please.

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  <div className={classes.root}>
+    {entries.map((entry) =>
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.tsx]: Please remove redundant `()` around `entry`.

File: src/CompletedTimeEntries/CompletedTimeEntries.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+import { action } from "@storybook/addon-actions";
+import { text } from "@storybook/addon-knobs";
+
+import { CompletedTimeEntries } from ".";
+
+export default { title: "CompletedTimeEntries" };
+
+export const initialState = () => (
+  <CompletedTimeEntries entries={[]}/>
+);
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.stories.tsx]: 1. Please rename the current story to `emptyState`.
2. Add story like `initialState` that contains some entries.
3. Use `knobs` to provide the ability to modify props in Storybook.

File: src/TimeEntries/TimeEntries.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import React from "react";
+
+import { TimeEntries } from ".";
+import { TimeEntryI } from "../TimeEntry.model";
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.stories.tsx]: I would add a style guide rule: sort app-level imports from general to specific. In your case, two lines should be swapped according this rule. Please don't forget to update README.

File: src/TimeEntries/TimeEntries.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+export interface TimeEntriesProps {
+  entries: TimeEntryI[];
+}
+
+interface Props
+  extends TimeEntriesProps,
+    WithStyles<typeof styles>,
+    WithTranslation {}
+
+const styles = ({ palette }: Theme) =>
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.tsx]: Styles are redundant, please remove.

File: src/TimeEntries/TimeEntries.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      color: palette.primary.main
+    }
+  });
+
+// prettier-ignore
+export const TimeEntries: ComponentType<TimeEntriesProps> = compose<Props, TimeEntriesProps>(
+   withTranslation(),
+   withStyles(styles, {withTheme: true, name: 'RTTSTimeEntries'})
+)(({classes, entries}) => {
+  return (
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.tsx]: Please use ES6 shorthand return syntax. I would also suggest adding this to the style guide.

File: src/TimeEntries/TimeEntries.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+export default { title: "TimeEntries" };
+
+export const emptyState = () => <TimeEntries entries={[]} />;
+
+const timeEntries: TimeEntryI[] = JSON.parse(
+  `[{"aspect":"aspect 1","project":"project 1","type":"type 1","task":"task 1"},{"aspect":"aspect 2","project":"project 2","type":"type 2","task":"task 2"},{"aspect":"aspect 3","project":"project 3","type":"type 3","task":"task 3"},{"aspect":"aspect 4","project":"project 4","type":"type 4","task":"task 4"},{"aspect":"aspect 5","project":"project 5","type":"type 5","task":"task 5"},{"aspect":"aspect 6","project":"project 6","type":"type 6","task":"task 6"}]`
+);
+
+export const listState = () => <TimeEntries entries={timeEntries} />;
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.stories.tsx]: Please use addon knobs.

File: src/TimeEntries/TimeEntries.tsx [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+     <div className={classes.root}>
+       {entries.length > 0 ?
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.tsx]: Please use `ramda`'s `isEmpty()`. In your case, you would need to write helper function `isNotEmpty` and place it in `src/utils.ts`.

File: src/TimeEntries/TimeEntries.tsx [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+         (
+            <List component="nav" aria-label="time entries">
+              <CompletedTimeEntries entries={entries}/>
+            </List>
+         ) : (
+            <div>It's nice time to begin.</div>
+          )
```
Comment (dezoxel) [inline src/TimeEntries/TimeEntries.tsx]: Such formatting issue (one redundant space) let me think that we need to check correctness of prettier pre-commit hook. I'll added it to the tech debt.

File: src/TimeEntry.model.ts [hunk 14]
```diff
@@ -0,0 +1,6 @@
+export interface TimeEntryI {
+  aspect: string;
+  project: string;
+  task: string;
+  type: string;
+}
```
Comment (dezoxel) [inline src/TimeEntry.model.ts]: Do we have an understanding what other models we can place to `TimeEntry.model` module? Maybe we have to make it more abstract in order to be able to add other models without creating the new module in the future?

File: src/TimeEntry/TimeEntry.stories.tsx [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+export default { title: "TimeEntry" };
+
+const mockEntry = {
+  aspect: "aspect",
+  project: "project",
+  task: "task",
+  type: "type"
+};
+
+export const initialState = () => <TimeEntry entry={mockEntry} />;
```
Comment (dezoxel) [inline src/TimeEntry/TimeEntry.stories.tsx]: I don't think we need to call it with prefix `mock`, it is redundant. Let's leave it just `entry`.

File: src/TimeEntry/TimeEntry.tsx [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+  extends TimeEntryProps,
+    WithStyles<typeof styles>,
+    WithTranslation {}
+
+const styles = ({ palette }: Theme) =>
+  createStyles({
+    root: {
+      color: palette.primary.main
+    }
+  });
```
Comment (dezoxel) [inline src/TimeEntry/TimeEntry.tsx]: Styles are redundant, please remove.

File: src/App/App.tsx [hunk 17]
```diff
@@ -20,6 +17,8 @@ import { compose } from "recompose";
 
 import { init as initI18N } from "../i18n";
 import { createTKey } from "../utils";
+import { TimeEntries } from "../TimeEntries";
+import { TimeEntryI } from '../TimeEntry.model';
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/App/App.tsx]: Done

File: src/App/App.tsx (hunk 1 above)
Comment (Sumragen) [reply src/App/App.tsx]: Done

File: src/TimeEntry.model.ts (hunk 14 above)
Comment (Sumragen) [reply src/TimeEntry.model.ts]: Done

File: src/TimeEntries/TimeEntries.stories.tsx (hunk 11 above)
Comment (Sumragen) [reply src/TimeEntries/TimeEntries.stories.tsx]: I didn't find the way how two 'mock' arrays using knob addon. It seems like there is no ability to use generic arrays

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface CompletedTimeEntriesProps {
+  entries: TimeEntryModel[];
+}
+
+interface Props extends CompletedTimeEntriesProps, WithTranslation {}
+
+// prettier-ignore
+export const CompletedTimeEntries: ComponentType<CompletedTimeEntriesProps> = compose<Props, CompletedTimeEntriesProps>(
+   withTranslation()
```
Comment (dezoxel) [inline src/CompletedTimeEntries/CompletedTimeEntries.tsx]: Translations redundant

File: src/TimeEntries/TimeEntries.tsx [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+
+export interface TimeEntriesProps {
+  entries: TimeEntryModel[];
+}
+
+interface Props extends TimeEntriesProps, WithTranslation {}
+
+// prettier-ignore
+export const TimeEntries: ComponentType<TimeEntriesProps> = compose<Props, TimeEntriesProps>(
+   withTranslation()
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntries/TimeEntries.tsx]: Agree. Removed

File: src/TimeEntry/TimeEntry.stories.tsx [hunk 20]
```diff
@@ -0,0 +1,21 @@
+import React from "react";
+
+import { TimeEntry } from ".";
+import { object, text, withKnobs } from '@storybook/addon-knobs';
+import { addDecorator } from '@storybook/react';
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.stories.tsx]: Yes, you are right. Moved TimeEntry import to separate line

File: src/TimeEntry/TimeEntry.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+
+addDecorator(withKnobs);
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.stories.tsx]: Done. Moved to config.ts file

File: src/TimeEntry/TimeEntry.stories.tsx [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+
+export default { title: "TimeEntry" };
+
+const entry = {
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.stories.tsx]: nice point. Specified

File: src/TimeEntry/TimeEntry.stories.tsx [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+  aspect: "aspect",
+  project: "project",
+  task: "task",
+  type: "type"
+};
+
+export const entryState = () => {
+  const knobEntry = object('TimeEntryModel', entry);
+  return (<TimeEntry entry={knobEntry} />);
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.stories.tsx]: Correct! Thanks. Done

File: src/TimeEntry/TimeEntry.tsx [hunk 24]
```diff
@@ -0,0 +1,30 @@
...
+
+export interface TimeEntryProps {
+  entry: TimeEntryModel;
+}
+
+interface Props extends TimeEntryProps, WithTranslation {}
+
+// prettier-ignore
+export const TimeEntry: ComponentType<TimeEntryProps> = compose<Props, TimeEntryProps>(
+   withTranslation()
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.tsx]: Correct. Removed

File: src/utils.spec.ts [hunk 25]
```diff
@@ -8,4 +8,42 @@ describe("Utils", () => {
...
     });
   });
+
+  describe("isNotEmpty", () => {
+    it("returns true on string", () => {
+      const value = "test_string_value";
+
+      expect(isNotEmpty(value)).toBe(true);
+    });
+    it("returns true on filled object", () => {
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/utils.spec.ts]: Done. Added.

File: src/utils.ts [hunk 26]
```diff
@@ -1,2 +1,4 @@
+import {isEmpty} from 'ramda';
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply src/utils.ts]: Done. Added

File: README.md [hunk 27]
```diff
@@ -25,6 +25,29 @@ Why?: Better readability.
...
+2. Use shorthand ES6 return syntax
+
+```javascript
+/* Bad */
+export const component = () => {
+    return (
+        <div>
+            content
+        </div>
+    )
```
[the remark this answers is not in the document]
Comment (Sumragen) [reply README.md]: Done

File: src/CompletedTimeEntries/CompletedTimeEntries.tsx (hunk 18 above)
Comment (Sumragen) [reply src/CompletedTimeEntries/CompletedTimeEntries.tsx]: Agree. Removed

File: README.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (dezoxel) [reply README.md]: I can still see 4 spaces on the first indent level

File: README.md [hunk 28]
```diff
@@ -25,6 +25,29 @@ Why?: Better readability.
...
+
+```
+
+```javascript
+/* Good */
+export const component = () => (
+   <div>
+        content
+   </div>
+);
```
Comment (dezoxel) [inline README.md]: Please 4 spaces.

File: src/TimeEntry/TimeEntry.tsx [hunk 29]
```diff
@@ -0,0 +1,24 @@
...
+
+import { TimeEntryModel } from "../models";
+
+export interface TimeEntryProps {
+  entry: TimeEntryModel;
+}
+
+// prettier-ignore
+export const TimeEntry: ComponentType<TimeEntryProps> = ({entry}: TimeEntryProps) => (
+   <>
```
Comment (dezoxel) [inline src/TimeEntry/TimeEntry.tsx]: No need `<>`

File: README.md (hunk 28 above)
Comment (Sumragen) [reply README.md]: Now there only 2 spaces on each level

File: src/TimeEntry/TimeEntry.tsx (hunk 29 above)
Comment (Sumragen) [reply src/TimeEntry/TimeEntry.tsx]: Yeah, you are right! Thanks

**Pull request merged**